### PR TITLE
Feature/Sidecar communication

### LINKED ISSUE #12: Implement the HTTP Service to expose the state to other sidecar operators and execute backup process.

### PR DESCRIPTION
**Pull request by iGxnon**
Base on #8, #3

Close #12 

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

    This PR aims to establish the foundational structure of the sidecar communication mechanism.

* what changes does this pull request make?

    - Implement the basic structure of the sidecar communication mechanism.
    - Provide the trait of backup
        - Implement the PV backup
    - Implement the backup route
    - Expose the state to other sidecar operators

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

    No

### CODE REVIEW

File: sidecar/src/backup/pv.rs [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+            dst.write_all(&item.blob).await?;
+        }
+        debug!(
+            "backup snapshot file: {filename}, size: {} MB",
+            size.div(0x0010_0000)
+        );
+        Ok(())
+    }
+
+    async fn load(&self, metadata: &Metadata) -> Result<PathBuf> {
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply sidecar/src/backup/pv.rs:66]: There might be errors in the implementation of other `Provider`s during `Provider::load`. For example, the `S3` load implementation should involve downloading the remote snapshot to the local filesystem, which could potentially encounter network issues. Here, the focus is on ensuring the consistency of the `Provider::load` signature.

File: sidecar/src/backup/pv.rs [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+            let Some(got) = src.next().await else {
+                return Err(anyhow!("get the item from source failed"))
+            };
+            item = got?;
+            size.add_assign(item.blob.len());
+            dst.write_all(&item.blob).await?;
+        }
+        debug!(
+            "backup snapshot file: {filename}, size: {} MB",
+            size.div(0x0010_0000)
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply sidecar/src/backup/pv.rs]: done, now we print how many KBs here

File: sidecar/src/operator.rs [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+
+        tokio::pin!(forceful_shutdown);
+        tokio::pin!(web_server);
+
+        let mut controller_shutdown = false;
+        let mut web_server_shutdown = false;
+
+        #[allow(clippy::integer_arithmetic)] // this error originates in the macro `tokio::select`
+        loop {
+            tokio::select! {
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply sidecar/src/operator.rs:85]: done, I'll fix `operator-k8s` in the following PRs

File: sidecar/src/controller.rs [hunk 4]
```diff
@@ -20,7 +20,7 @@ pub(crate) struct Controller {
     /// Check interval
     check_interval: Interval,
     /// graceful shutdown signal
-    graceful_shutdown: EventListener,
+    graceful_shutdown: Receiver<()>,
```
Comment (Phoenix500526) [inline sidecar/src/controller.rs:23]: How come you replace `EventListener` with `Receiver`? Is it better?

Comment (iGxnon) [reply sidecar/src/controller.rs:23]: The reason similar to [xline's graceful shutdown](https://github.com/xline-kv/Xline/pull/396#issuecomment-1649766700).
When running multiple tasks in parallel using `tokio::spawn`, it's essential to have a "callback" mechanism to signal the `run` once all tasks have finished.
Implementing it using an `Receiver` is cleaner than `EventListener`.

**Review by Phoenix500526 — DISMISSED** [review]
LGTM

File: sidecar/src/backup/pv.rs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        let mut item = item?;
+        let mut size = item.blob.len();
+        dst.write_all(&item.blob).await?;
+        while item.remaining_bytes > 0 {
+            let Some(got) = src.next().await else {
+                return Err(anyhow!("get the item from source failed"))
+            };
+            item = got?;
+            size.add_assign(item.blob.len());
+            dst.write_all(&item.blob).await?;
```
Comment (bsbds) [inline sidecar/src/backup/pv.rs:57]: I think this should be append to file. And we can write a snapshot to memory and flush it to disk at one time for better performance.

File: sidecar/src/main.rs [hunk 6]
```diff
@@ -212,13 +220,12 @@ fn parse_backup_type(value: &str) -> Result<Backup, String> {
         "s3" => {
             if items.len() != 2 {
```
Comment (bsbds) [inline sidecar/src/main.rs:221]: This should also change to `1`

File: sidecar/src/backup/pv.rs (hunk 5 above)
Comment (iGxnon) [reply sidecar/src/backup/pv.rs]: done

File: sidecar/src/main.rs (hunk 6 above)
Comment (iGxnon) [reply sidecar/src/main.rs]: done

**Review by Phoenix500526 — APPROVED** [review]

**Review by bsbds — APPROVED** [review]

**Pull request merged**